File: Python/String/33_rearrange_characters.py
"""
Idea:
put the highest frequency characters first & using priority queue to put all elements in order of their frequency
Inserting element & removing it from priority queue, temporarily. appending next highest frequency element, appending the removed element.
"""
import heapq


class Solution:
    def rearrangeString(self, S):
        # Creating a dictionary to count the occurrences of each character in the string
        count = {}
        for c in S:
            count[c] = 0

        # Counting the occurrences of each character in the string
        for c in S:
            count[c] += 1

        # Creating a heap and storing the characters with their negative counts in the heap
        heap = []
        for (key, value) in count.items():
            heap.append([-value, key])

        # Converting the list into a min-heap
        heapq.heapify(heap)

        # Initializing a variable to keep track of the previous character and its count
        prev = [0, '#']

        # Initializing an empty string to store the rearranged string
        ans = ""

        # Repeatedly extracting the character with the highest count from the heap and appending it to the result string
        while len(heap) != 0:
            key2 = heapq.heappop(heap)
            ans += key2[1]

            # If the count of the previous character is still negative, push it back into the heap
            if -prev[0] > 0:
                heapq.heappush(heap, prev)
            key2[0] += 1
            prev = key2

        # Checking if the rearranged string has the same length as the input string
        if len(ans) == len(S):
            return ans
        else:
            return "-1"


if __name__ == '__main__':
    t = int(input())
    for _ in range(t):
        str1 = input()
        solObj = Solution()
        str2 = solObj.rearrangeString(str1)
        if str2 == '-1':
            print(0)
        elif sorted(str1) != sorted(str2):
            print(0)
        else:
            for i in range(len(str2) - 1):
                if str2[i] == str2[i + 1]:
                    print(0)
                    break
            else:
                print(1)

File: C++/String/33_rearrange_characters.cpp
/*
Idea:
put the highest frequency characters first & using priority queue to put all elements in order of their frequency
Inserting element & removing it from priority queue, temporarily. appending next highest frequency element, appending the removed element.
*/


#include<bits/stdc++.h>
using namespace std;
const int MAX_CHAR = 26;


class Solution
{
public:
    struct Key
    {
        int freq;
        char ch;

        // Function for priority_queue to store Key  according to freq
        bool operator<(const Key& k) const
        {
            return freq < k.freq;
        }
    };
    string rearrangeString(string str)
    {
        int n = str.size();

        int count[MAX_CHAR] = { 0 };
        for (int i = 0;i < n;++i)
            count[str[i] - 'a']++;

        priority_queue<Key> pq;

        //Insert all characters with their frequencies into a priority_queue
        for (char c = 'a';c <= 'z';++c)
        {
            int val = c - 'a';
            if (count[val])
                pq.push(Key{ count[val],c });
        }
        string ans = "";
        Key prev{ -1,'#' }; // work as the previous visited element

        while (!pq.empty())
        {
            Key k = pq.top();
            pq.pop();
            str = str + k.ch; // appending top value of priority queue

            if (prev.freq > 0)
                // IF frequency of previous character is less  than zero that means it is useless, 
                // so no need to insert in priority queue again
                pq.push(prev);

            k.freq--;
            prev = k;
        }

        // if resultant length is not equal to original, return false
        if (n != ans.size())
            return "-1";
        else
            return ans;
    }

};


int main()
{
    int t;
    cin >> t;
    while (t--)
    {
        string str;
        cin >> str;
        Solution ob;
        string str1 = ob.rearrangeString(str);
        int flag = 1;
        int c[26] = { 0 };
        for (int i = 0; i < str.length(); i++)
            c[str[i] - 'a'] += 1;
        int f = 0;
        int x = (str.length() + 1) / 2;
        for (int i = 0; i < 26; i++)
        {
            if (c[i] > x)
                f = 1;
        }
        if (f)
        {
            if (str1 == "-1")
                cout << 0 << endl;
            else
                cout << 1 << endl;
        }
        else
        {
            int a[26] = { 0 };
            int b[26] = { 0 };
            for (int i = 0; i < str.length(); i++)
                a[str[i] - 'a'] += 1;
            for (int i = 0; i < str1.length(); i++)
                b[str1[i] - 'a'] += 1;

            for (int i = 0; i < 26; i++)
                if (a[i] != b[i])
                    flag = 0;

            for (int i = 0;i < str1.length();i++)
            {
                if (i > 0)
                    if (str1[i - 1] == str1[i])
                        flag = 0;
            }
            if (flag == 1)
                cout << "1\n";
            else
                cout << "0\n";
        }
    }
    return 0;
}
